### PR TITLE
upgrade: controller-runtime and code change accordingly

### PR DESCRIPTION
**Pull request by zdtsw**
## Description
**split origin PR** https://github.com/opendatahub-io/opendatahub-operator/pull/1048 **into 2 parts.**
this PR is part I, focusing on upgrade controller-runtime and clenaup go packages.
details:
- the current controller-runtime 0.14.6 does not support new features:
- uplift controller-runtime to v0.17.5 which is the last support go 1.21
- update Watches() to match 0.17.5
- webhook in the new 0.16.3 does not use method InjectDecoder InjectClient, updates
- remove MWC since we are not using it for now
- cleanup old go.mod, remove unnecessary replacement
- some formatting for SetupWithManager()
- add missing "get" on console

## How Has This Been Tested?

local build: quay.io/wenzhou/opendatahub-operator-catalog:v2.17.2-60

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: config/webhook/service.yaml [hunk 1]
```diff
@@ -1,18 +1,13 @@
...
-    app.kubernetes.io/instance: webhook-service
-    app.kubernetes.io/component: webhook
     app.kubernetes.io/created-by: opendatahub-operator
     app.kubernetes.io/part-of: opendatahub-operator
     app.kubernetes.io/managed-by: kustomize
   name: webhook-service
-  namespace: system
   annotations:
-    service.beta.openshift.io/serving-cert-secret-name: opendatahub-operator-controller-webhook-cert
+    service.beta.openshift.io/serving-cert-secret-name: opendatahub-operator-webhook-cert
```
Comment (ykaliuta) [inline config/webhook/service.yaml:10]: Are these changes related to the update?

And more general: can I take webhook changes as a separate PR? They do not require the update, can go now already and it will be easier then to take the patch downstream.

Comment (zdtsw) [reply config/webhook/service.yaml:10]: i can leave the changes in bundle/ + config/webhook/ out from this PR
but the one in the controllers/webhook/ will be needed for 0.17.5 to work

Comment (ykaliuta) [reply config/webhook/service.yaml:10]: Yes, but webhook initialization changes do not require updated runtime and can be submitted right now, don't they?

Comment (zdtsw) [reply config/webhook/service.yaml:10]: if i understand you correctly:
you want to have another PR only for the webhook decoder part in ODH first, and you want to get it merge into downstream without introduce the upgrade runtime in downstream.

Comment (ykaliuta) [reply config/webhook/service.yaml:10]: > if i understand you correctly: you want to have another PR only for the webhook decoder part in ODH first, and you want to get it merge into downstream without introduce the upgrade runtime in downstream.

I want webhook changes in a separate PR, it may be more general with both cleanup in names and manifests (if you want it) and decoder part. To track it for backporting separately from the upgrade.

Comment (zdtsw) [reply config/webhook/service.yaml:10]: ok, i have no objection on this

Comment (zdtsw) [reply config/webhook/service.yaml]: i have narrow the changes in this PR, explicitly removed some part of webhook.
but due to the fact with a newer version or controller-runtime the old implementation on webhook has been changed, in this case: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.6/pkg/webhook/admission#InjectDecoderInto is gone.

**Comment by ykaliuta** [comment]
There are some tiny unrelated changes (some formatting and error texts),  but I can put my ACK. Just will wait if @bartoszmajsak has comments.

File: main.go [hunk 2]
```diff
@@ -147,7 +153,10 @@ func main() { //nolint:funlen
 		os.Exit(1)
 	}
 
-	(&webhook.OpenDataHubWebhook{}).SetupWithManager(mgr)
+	(&webhook.OpenDataHubWebhook{
+		Client:  mgr.GetClient(),
```
[the remark this answers is not in the document]
Comment (ykaliuta) [reply main.go:157]: Good to think about it, but do not see an issue. It lives in the same binary, does not reset the cache, just makes potentially a couple of lookups on creation/deletion of DSCI/DSC. But I can miss the whole picture.

[the remark this answers is not in the document]
Comment (ykaliuta) [reply main.go:157]: There is probably some race window on deletion (deleted DSC not yet invalidated in the cache and webhook blocks DSC deletion) but IIUC it's pretty small and does not cause problems in practice, delete request is not executed instantly either.

File: main.go [hunk 3]
```diff
@@ -123,10 +125,14 @@ func main() { //nolint:funlen
...
-		Scheme:                 scheme,
-		MetricsBindAddress:     metricsAddr,
-		Port:                   9443,
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{ // single pod does not need to have LeaderElection
+		Scheme:  scheme,
+		Metrics: ctrlmetrics.Options{BindAddress: metricsAddr},
+		WebhookServer: ctrlwebhook.NewServer(ctrlwebhook.Options{
+			Port: 9443,
+			// TLSOpts: , // TODO: do we need tls for webhook
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply main.go:134]: i think we have some jira already have items on these across the project, i only put a TODO here as a marker

File: go.mod [hunk 4]
```diff
@@ -58,55 +58,47 @@ require (
...
-	github.com/mattn/go-sqlite3 => github.com/mattn/go-sqlite3 v1.14.18
-	github.com/tektoncd/pipeline => github.com/tektoncd/pipeline v0.12.0
-	go.uber.org/multierr => go.uber.org/multierr v1.6.0
-	google.golang.org/grpc => google.golang.org/grpc v1.63.2
-	k8s.io/api => k8s.io/api v0.26.0
-	k8s.io/client-go => k8s.io/client-go v0.26.0
-	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20230113010536-ace3e4a52ca1
-	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.14.6
+	k8s.io/api => k8s.io/api v0.28.4
+	k8s.io/client-go => k8s.io/client-go v0.28.4
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply go.mod:101]: i will need to check these two, with a different ticket, just a quick look it will break scheme and flowcontrol

File: controllers/dscinitialization/dscinitialization_controller.go [hunk 5]
```diff
@@ -381,18 +420,16 @@ func (r *DSCInitializationReconciler) watchMonitoringSecretResource(a client.Obj
...
-		}
-		if len(instanceList.Items) == 0 && !upgrade.HasDeleteConfigMap(ctx, r.Client) {
-			r.Log.Info("Found no DSC instance in cluster but not in uninstalltion process, reset monitoring stack config")
-
-			return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: "backup"}}}
-		}
+func (r *DSCInitializationReconciler) watchDSCResource(ctx context.Context) []reconcile.Request {
+	instanceList := &dscv1.DataScienceClusterList{}
+	if err := r.Client.List(ctx, instanceList); err != nil {
+		// do not handle if cannot get list
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/dscinitialization/dscinitialization_controller.go:426]: updated.

File: main.go [hunk 6]
```diff
@@ -160,11 +169,11 @@ func main() { //nolint:funlen
 		os.Exit(1)
 	}
 
-	if err = (&datascienceclustercontrollers.DataScienceClusterReconciler{
+	if err = (&dsccontr.DataScienceClusterReconciler{
 		Client: mgr.GetClient(),
 		Scheme: mgr.GetScheme(),
 		Log:    logger.LogWithLevel(ctrl.Log.WithName(operatorName).WithName("controllers").WithName("DataScienceCluster"), logmode),
-		DataScienceCluster: &datascienceclustercontrollers.DataScienceClusterConfig{
+		DataScienceCluster: &dsccontr.DataScienceClusterConfig{
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply main.go:176]: follow: https://github.com/opendatahub-io/opendatahub-operator/pull/1193

File: main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply main.go:157]: Fair enough, it's preserving the behaviour. We can optimize later if needed

File: controllers/dscinitialization/dscinitialization_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply controllers/dscinitialization/dscinitialization_controller.go:426]: Maybe add an error message too, otherwise it's not really actionable - we are loosing information WHY it is failing.

[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/dscinitialization/dscinitialization_controller.go:426]: change from  info  to error

**Review by bartoszmajsak — COMMENTED** [review]
That's a long-awaited upgrade :) Thanks for working on it. I will let someone from the team to formally approve it.

**Review by ykaliuta — APPROVED** [review]
